runtime/v2/shim/util.go

```diff
@@ -116,41 +117,60 @@ func Connect(address string, d func(string, time.Duration) (net.Conn, error)) (n
 }
 
 // WritePidFile writes a pid file atomically
-func WritePidFile(path string, pid int) error {
-	path, err := filepath.Abs(path)
+func WritePidFile(path string, pid int) (err error) {
+	path, err = filepath.Abs(path)
 	if err != nil {
 		return err
 	}
-	tempPath := filepath.Join(filepath.Dir(path), fmt.Sprintf(".%s", filepath.Base(path)))
-	f, err := os.OpenFile(tempPath, os.O_RDWR|os.O_CREATE|os.O_EXCL|os.O_SYNC, 0666)
+
+	f, err := os.CreateTemp(filepath.Dir(path), "."+filepath.Base(path))
 	if err != nil {
 		return err
 	}
-	_, err = fmt.Fprintf(f, "%d", pid)
-	f.Close()
-	if err != nil {
+	defer func() {
+		f.Close()
+		if err == nil {
+			err = os.Rename(f.Name(), path)
+		}
+	}()
+
+	if _, err := f.WriteString(strconv.Itoa(pid)); err != nil {
 		return err
 	}
-	return os.Rename(tempPath, path)
+
+	if err := f.Chmod(0666); err != nil {
+		return err
+	}
+
+	return nil
 }
 
 // WriteAddress writes a address file atomically
-func WriteAddress(path, address string) error {
-	path, err := filepath.Abs(path)
+func WriteAddress(path, address string) (err error) {
+	path, err = filepath.Abs(path)
 	if err != nil {
 		return err
 	}
-	tempPath := filepath.Join(filepath.Dir(path), fmt.Sprintf(".%s", filepath.Base(path)))
-	f, err := os.OpenFile(tempPath, os.O_RDWR|os.O_CREATE|os.O_EXCL|os.O_SYNC, 0666)
+	f, err := os.CreateTemp(filepath.Dir(path), "."+filepath.Base(path))
```

Comment:
Sorry. I don't get your point here. The `tempPath` has temp prefix, but it doesn't mean it is named randomly.
The `WriteAddress` is to create temp file named `.X` and then rename it to `X`, just in case that the X content is corrupted after the node's power outage or out of disk space. 

I can't tell any optimization from your patch. The original `OpenFile` is very straightforward. But this patch will pick up a unused name and open it with 0600 permission. And then it files other syscall to chmod to 0666....
One syscall -> More than two syscalls.

Please point it out if I miss somethings. Thanks

Reply:
Oop, I never submitted my review besides my stray comment. I mostly have the same feedback here, I don't get what this is optimizing. Do you have numbers to share/could you put in the description what the optimization is?

Reply:
1. use `os.CreateTemp()` to ensure that the temporary file name does not conflict with the existing file name, more secure and reliable

2. use `f.Write()` to write the pid instead of `fmt.Fprintf()` because the former does not require string formatting and is therefore faster

3. use `defer f.Close()` to guarantee that the file will be closed if opened in any case

Reply:
Thanks for the quick reply.

Basically, if `Write{PidFile,Address}` fails, it won't retry and the bundle will be deleted. So, it won't be conflict issue. 

For the part 2, it can be accepted. Thanks

And for part 3 , it's based on the statements. Since the `Close` after write, it is safe (at least to me).

Reply:
My idea is that temporary data may be written to an existing file, resulting in the original file data being copied, although the probability is not high.

Reply:
It's based on the usecase. In shim, the temporary file is `.X`. The prefix `.` means temporary.
But in common case, your change is good. We've already provided it as AtomicWriteFile. I think we can handle the close in defer.

https://github.com/containerd/continuity/blob/v0.3.0/ioutils.go#L28

Reply:
got it
(Sorry, I am a containerd novice and do not know this. I am currently learning containerd. I hope I can participate in community development through some of my own ideas.😀

Reply:
Neven mind! Welcome to contribute to container project. The code review can recall memory for the code. :p